### PR TITLE
Support (BPF) filter in sniff() with offline parameter set

### LINKED ISSUE #355: The sniff() filter parameter is not documented
The `sniff()` filter parameter is not documented with `help()`.

### LINKED ISSUE #393: Scapy ignoring filter argument to sniff() function
Reopening bitbucket issue nr. 747, https://bitbucket.org/secdev/scapy/issues/747/scapy-ignoring-filter-argument-to-sniff

Scapy seems to ignore the `filter` argument to `sniff()` in my installation, scapy 2.3.2 on debian.

I believe I can exclude tcpdump from being the issue as I moved it out of the way for this test, using an offline pcap file:
```py
In [1]: from scapy.all import *
WARNING: Failed to execute tcpdump. Check it is installed and in the PATH

In [2]: pkts = sniff(offline="/tmp/some.pcap", filter="udp and host 7.7.7.7", prn=lambda x: x.summary())
Ether / IP / TCP 192.5.XXX.XXX:33658 > 85.XX.XX.XX:8020 S
Ether / IP / TCP 85.XX.XX.XX:8020 > 192.5.XXX.XXX:33658 SA
...
```
As can be seen, the `udp` filter is ignored while listing all kinds of other packets (tcp, icmp etc., any source or destination).

### PR DESCRIPTION
**Pull request by p-l-**
Fixes #393
Also, fixes #355

`filter=` was not documented in `sniff()` (#355) because it is only passed to the socket. For the same reason, it was not supported with `offline=`, because no socket is used in that case.

Since there is no way (apart from writing a BPF filter parser in Scapy) to filter a PCAP, this patch uses `tcpdump`.

### CODE REVIEW

**Comment by nyov** [comment]
This works for me! Awesome work.
So I see the issue was only because I tested an offline pcap file. Oh well.

Couldn't handle gzipped pcap, but then that seems tcpdump's limitation.
Thanks!

File: scapy/arch/windows/compatibility.py [hunk 1]
```diff
@@ -186,8 +187,25 @@ def sniff(count=0, store=1, offline=None, prn = None, lfilter=None, L2socket=Non
...
+                else:
+                    s = PcapReader(
+                        subprocess.Popen(
+                            [conf.prog.tcpdump, "-r", "-", "-w", "-", flt],
+                            stdin=offline,
+                            stdout=subprocess.PIPE
+                        ).stdout
+                    )
+            else:
+                s = PcapReader(offline)
```
Comment (nyov) [inline scapy/arch/windows/compatibility.py]: FYI. Indent level here seems wrong.

Comment (p-l-) [inline scapy/arch/windows/compatibility.py]: Nice catch!

**Comment by p-l-** [comment]
Thanks for the tests and review @nyov!

**Comment by guedou** [comment]
Looks perfect !

You could add the following unit tests to `regression.uts`. This option is a good opportunity to test sniff. What do you think ?

```python
import tempfile
fd, filename = tempfile.mkstemp()
wrpcap(filename, Ether(dst="ff:ff:ff:ff:ff")/IP(dst=["1.2.3.4", "5.6.7.8"])/ICMP())
len(sniff(offline=filename, filter="host 5.6.7.8")) == 1
```

**Comment by p-l-** [comment]
@guedou @gpotter2 any idea on how I could easily install `tcpdump` in AppVeyor? So far I have disabled tests requiring `tcpdump` (in AppVeyor and in Travis-CI when `tcpdump` is not in `$PATH`).

**Comment by gpotter2** [comment]
I'm performing some tests to see how you could do that...

**Comment by p-l-** [comment]
Thanks @gpotter2. If/when you find a solution, could you submit a PR?
In the meantime, if the tests pass and are OK for you @guedou, I'd say good to merge.

**Comment by gpotter2** [comment]
@p-l- 
On my windows computer, adding this:
```
  - ps: wget http://www.winpcap.org/windump/install/bin/windump_3_9_5/WinDump.exe -UseBasicParsing -OutFile C:\Windows\System32\Npcap\windump.exe
```
between
`  - choco install -y nmap`
and
`  - refreshenv`

in appveyor.yml did the trick... But i'm not sure that it will work on Appveyor...

**Comment by p-l-** [comment]
For some reason that does not work.

@gpotter2 have you been able to use `sniff(offline=..., filter=...` on Windows?

I suspect `-w -` is not working here (should be however, according to [Windump manual](https://www.winpcap.org/windump/docs/manual.htm)).

**Comment by gpotter2** [comment]
Weirdo: i tried with the current scapy version:

I manually launched 

> wget http://www.winpcap.org/windump/install/bin/windump_3_9_5/WinDump.exe -UseBasicParsing -OutFile C:\Windows\System32\Npcap\windump.exe

with the admin rights on powershell. Then added:
`set PATH=C:/Windows/System32/Npcap/;%PATH%`
at the top of `run_scapy.bat`

Then i manually pasted the test:

http://pastebin.com/Gw55yMXa

And it worked :/

**Comment by gpotter2** [comment]
Investigation is in progress: the "magic" is empty when loading on Appveyor, which cause the crash. The file is readable + writable

**Comment by p-l-** [comment]
So when I run the tests, the temporary file from which I read contains (base64 encoded here):
```
1MOyoQIABAAAAAAAAAAAAP//AABlAAAAz8WsVm8qCgAoAAAAKAAAAEUAACgAAQAAQAZ8zX8AAAF/
AAABABQAUAAAAAAAAAAAUAIgAJF8AADPxaxWXy0KABwAAAAcAAAARQAAHAABAABAEXzOfwAAAX8A
AAEANQA1AAgBcs/FrFb5MAoAHAAAABwAAABFAAAcAAEAAEABfN5/AAABfwAAAQgA9/8AAAAA
```

According to `tcpdump` on my Linux machine, that's correct:
```
$ echo 1MOyoQIABAAAAAAAAAAAAP//AABlAAAAz8WsVm8qCgAoAAAAKAAAAEUAACgAAQAAQAZ8zX8AAAF/AAABABQAUAAAAAAAAAAAUAIgAJF8AADPxaxWXy0KABwAAAAcAAAARQAAHAABAABAEXzOfwAAAX8AAAEANQA1AAgBcs/FrFb5MAoAHAAAABwAAABFAAAcAAEAAEABfN5/AAABfwAAAQgA9/8AAAAA \
>  | base64 -d | tcpdump -n -r -
reading from file -, link-type RAW (Raw IP)
15:16:47.666223 IP 127.0.0.1.20 > 127.0.0.1.80: Flags [S], seq 0, win 8192, length 0
15:16:47.666975 IP 127.0.0.1.53 > 127.0.0.1.53: [|domain]
15:16:47.667897 IP 127.0.0.1 > 127.0.0.1: ICMP echo request, id 0, seq 0, length 8
```

But when, in AppVeyor, I run with subprocess `['C:\\Windows\\System32\\Npcap\\windump.exe', '-r', 'c:\\users\\appveyor\\appdata\\local\\temp\\tmp8rdgva']`, I get both stdout and stderr empty. So for some reason `windump` is not able to read the PCAP file. Is that because of the link-type (Raw IP)? What I don't understand is why I don't even get an error message.

**Comment by p-l-** [comment]
I have tried to add an `Ether()` layer to change the link-layer from RAW (Raw IP) to EN10MB (Ethernet), same results. So to sum up:
* Raw IP link layer is not the cause
* Using STDIO (`-r -` and/or `-w -`) is not the cause
* The file on the disk is correct (it's a PCAP file that can be read without any problem on Linux)
* Windump does not read the file (`windump -r <filename>` produces no output or error)

Maybe Windump is not correctly installed or needs something else? But in that case, wouldn't it produce at least an error?

**Comment by gpotter2** [comment]
I've performed some tests on AppVeyor, and windump.exe does exists in the path... I guess we're getting it output badly...

**Comment by p-l-** [comment]
@gpotter2 I have no doubt it exists in the PATH, I have a clear error when it was not installed. I don't have access to a Windows machine for now, so if you can, can you do that test for me (see my previous comment)

**Comment by gpotter2** [comment]
Sure

**Comment by gpotter2** [comment]
@p-l- 
```
Z:\>windump -r file.pcap
reading from file file.pcap, link-type RAW (Raw IP)
15:16:47.666223 IP ZALMAN.20 > ZALMAN.80: S 0:0(0) win 8192
15:16:47.666975 IP ZALMAN.53 > ZALMAN.53: [|domain]
15:16:47.667897 IP ZALMAN > ZALMAN: ICMP echo request, id 0, seq 0, length 8
```

**Comment by p-l-** [comment]
@gpotter2 thanks a lot. That's weird. Can you try something else for me?
```
python2 -c "from subprocess import Popen, PIPE; print subprocess.Popen(['windump', '-r', 'file.cap'], stdout=PIPE, stderr=PIPE).communicate()"
```

If that also works, the problem is probably that Windump is not correctly installed in AppVeyor (something is missing for example).

**Comment by gpotter2** [comment]
Weirdly, i cannot manage to select the file... I will always have:
```
Z:\>python -c "import subprocess; from subprocess import Popen, PIPE; print subprocess.Popen(['windump', '-r', 'file.cap'], stdout=PIPE, stderr=PIPE).communicate()"
('', 'windump: file.cap: No such file or directory\r\n')
```
```
Z:\>python -c "import subprocess; from subprocess import Popen, PIPE; print subprocess.Popen(['windump', '-r', 'Z:\file.cap'], stdout=PIPE, stderr=PIPE).communicate()"
('', 'windump: Z:/file.cap: No such file or directory\r\n')
```

```
Z:\>python -c "import subprocess; from subprocess import Popen, PIPE; print subprocess.Popen(['windump', '-r', './file.cap'], stdout=PIPE, stderr=PIPE).communicate()"
('', 'windump: Z:\file.cap: No such file or directory\r\n')
```

**Comment by p-l-** [comment]
@gpotter2 can you try `"Z:\\file.cap"`?

**Comment by gpotter2** [comment]
@p-l- 
```
C:\Users\gpotter\Desktop>python -c "import subprocess; from subprocess import Popen, PIPE; print subprocess.Popen(['windump', '-r', 'C:\Users\gpotter\Desktop\file.cap'], stdout=PIPE, stderr=PIPE).communicate()"
('', 'windump: C:\\Users\\gpotter\\Desktop\x0cile.cap: Invalid argument\r\n')
```

**Comment by gpotter2** [comment]
@p-l- You might be right:
I was assuming that @guedou's work aroung (installing winpcap-nmap) would install winpcap, but it's a modified version so as this says:
https://www.winpcap.org/windump/install/
The requirements won't be required i guess...

**Comment by p-l-** [comment]
@gpotter2 I think you should double each `\`, shouldn't you? so probably something like `python -c "import subprocess; from subprocess import Popen, PIPE; print subprocess.Popen(['windump', '-r', 'C:\\Users\\gpotter\\Desktop\\file.cap'], stdout=PIPE, stderr=PIPE).communicate()"`

**Comment by p-l-** [comment]
If that's because of Windump install I might just skip Windows tests requiring Windump for now...

**Comment by gpotter2** [comment]
@p-l- Well i've found a fix

**Comment by p-l-** [comment]
@gpotter2 Cool! What is it?

**Comment by gpotter2** [comment]
@p-i- see my comment above

**Comment by p-i-** [comment]
@p-l- see @gpotter2 's comment above ;)

**Comment by gpotter2** [comment]
Hu ? How many are they ?

**Comment by p-l-** [comment]
Thanks @p-i- and sorry for the notification ;)

**Comment by p-l-** [comment]
Thanks @gpotter2! AppVeyor tests are now OK (and I have canceled the other builds --- it takes a long time to start new MacOS tests on Travis-CI, queue "OSX for open source" has been full for 8 hours, and counting).

**Comment by guedou** [comment]
Thanks for the unit tests ! They are really useful.

I am glad to see that you find another way to install wpcap.dll. That's something that we might want to document.

**Comment by nohadrweesh** [comment]
I'm afraid I can't understand it well,can you provide an example of using tcdump to filter captured packets?

**Comment by gpotter2** [comment]
@nohadrweesh Please open a new issue if really needed, you should fix your sample as:
`pkts=sniff(offline='file.pcap',filter='tcp')`

**Pull request merged**